svc.go

```diff
@@ -0,0 +1,180 @@
+	ctx    context.Context
+	dialer host.Host
+	peers  map[peer.ID]struct{}
+	mx     sync.Mutex
+}
+
+func NewAutoNATService(ctx context.Context, h host.Host) (*AutoNATService, error) {
+	dialer, err := libp2p.New(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	as := &AutoNATService{
+		ctx:    ctx,
+		dialer: dialer,
+		peers:  make(map[peer.ID]struct{}),
+	}
+	h.SetStreamHandler(AutoNATProto, as.handleStream)
+
+	go as.resetPeers()
+
+	return as, nil
+}
+
+func (as *AutoNATService) handleStream(s inet.Stream) {
+	defer s.Close()
+
+	pid := s.Conn().RemotePeer()
+	log.Debugf("New stream from %s", pid.Pretty())
+
+	r := ggio.NewDelimitedReader(s, inet.MessageSizeMax)
+	w := ggio.NewDelimitedWriter(s)
+
+	var req pb.Message
+	var res pb.Message
+
+	err := r.ReadMsg(&req)
+	if err != nil {
+		s.Reset()
+		return
+	}
+
+	t := req.GetType()
+	if t != pb.Message_DIAL {
+		log.Debugf("Unexpected message from: %s", t.String())
+		s.Reset()
+		return
+	}
+
+	dr := as.handleDial(pid, req.GetDial().GetPeer())
+	res.Type = pb.Message_DIAL_RESPONSE.Enum()
+	res.DialResponse = dr
+
+	err = w.WriteMsg(&res)
+	if err != nil {
+		log.Debugf("Error writing response: %s", err.Error())
+		s.Reset()
+		return
+	}
+}
+
+func (as *AutoNATService) handleDial(p peer.ID, mpi *pb.Message_PeerInfo) *pb.Message_DialResponse {
+	if mpi == nil {
+		return newDialResponseError(pb.Message_E_BAD_REQUEST, "missing peer info")
+	}
+
+	mpid := mpi.GetId()
+	if mpid != nil {
+		mp, err := peer.IDFromBytes(mpid)
+		if err != nil {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "bad peer id")
+		}
+
+		if mp != p {
+			return newDialResponseError(pb.Message_E_BAD_REQUEST, "peer id mismatch")
+		}
+	}
+
+	addrs := make([]ma.Multiaddr, 0)
+	for _, maddr := range mpi.GetAddrs() {
+		addr, err := ma.NewMultiaddrBytes(maddr)
+		if err != nil {
+			log.Debugf("Error parsing multiaddr: %s", err.Error())
+			continue
+		}
+
+		// skip relay addresses
+		_, err = addr.ValueForProtocol(P_CIRCUIT)
+		if err == nil {
+			continue
+		}
+
+		// skip private network (unroutable) addresses
+		if !isPublicAddr(addr) {
+			continue
+		}
+
+		addrs = append(addrs, addr)
+	}
+
+	if len(addrs) == 0 {
+		return newDialResponseError(pb.Message_E_DIAL_ERROR, "no dialable addresses")
+	}
+
+	return as.doDial(pstore.PeerInfo{ID: p, Addrs: addrs})
+}
+
+func (as *AutoNATService) doDial(pi pstore.PeerInfo) *pb.Message_DialResponse {
+	// rate limit check
+	as.mx.Lock()
+	_, ok := as.peers[pi.ID]
+	if ok {
+		as.mx.Unlock()
+		return newDialResponseError(pb.Message_E_DIAL_REFUSED, "too many dials")
+	}
+	as.peers[pi.ID] = struct{}{}
+	as.mx.Unlock()
+
+	ctx, cancel := context.WithTimeout(as.ctx, 42*time.Second)
+	defer cancel()
+
+	err := as.dialer.Connect(ctx, pi)
+	if err != nil {
+		log.Debugf("error dialing %s: %s", pi.ID.Pretty(), err.Error())
+		// wait for the context to timeout to avoid leaking timing information
+		// this renders the service ineffective as a port scanner
+		select {
```

Comment:
You don't have to `select` here.
`<-ctx.Done` is enough.

Reply:
yes, indeed; habbit.